R/grid.R

```diff
@@ -0,0 +1,295 @@
+#' @param cols_width vector of strings
+#'
+#' @return string
+#' @details This is a helper function used in grid_template()
+#' \preformatted{
+#'   grid_container_css(
+#'     "'a a a' 'b b b'",
+#'     c("50\%", "50\%"),
+#'     c("100px", "2fr", "1fr")
+#'   )
+#' }
+#' returns
+#' \preformatted{
+#'   "display: grid;
+#'    height: 100\%;
+#'    grid-template-rows: 50\% 50\%;
+#'    grid-template-columns: 100px 2fr 1fr;
+#'    grid-template-areas: 'a a a' 'b b b';
+#'    {custom_style_grid_container}"
+#' }
+#'
+grid_container_css <- function(css_grid_template_areas, rows_height, cols_width) {
+  grid_container_styles <- c(
+    "display: grid",
+    "height: 100%",
+    glue::glue("grid-template-rows: {paste(rows_height, collapse = ' ')}"),
+    glue::glue("grid-template-columns: {paste(cols_width, collapse = ' ')}"),
+    glue::glue("grid-template-areas: {css_grid_template_areas}"),
+    "{custom_style_grid_container}" # To be rendered by glue::glue() later
+  )
+  paste(grid_container_styles, collapse = "; ")
+}
+
+#' Generate list of HTML div elements representing grid areas.
+#'
+#' @param area_names
+#'
+#' @return list of shiny::tags$div
+#'
+#' @details This is a helper function used in grid_template()
+#' \preformatted{
+#'   list_of_area_tags(c("header", "main", "footer"))
+#' }
+#' returns the following list:
+#' \preformatted{
+#'   [[1]] <div style="grid-area: header; {custom_style_grid_area_header}">{{ header }}</div>
+#'   [[2]] <div style="grid-area: main; {custom_style_grid_area_main}">{{ main }}</div>
+#'   [[3]] <div style="grid-area: footer; {custom_style_grid_area_footer}">{{ footer }}</div>
+#' }
+#'
+list_of_area_tags <- function(area_names) {
+  lapply(area_names,
+    function(name) shiny::tags$div(
+     style = as.character(glue::glue("grid-area: {name}; {{custom_style_grid_area_{name}}}")), paste("{{", name, "}}")
+    )
+  )
+}
+
+#' Define a template of a CSS grid
+#'
+#' @param default
+#' Template for desktop:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#' @param mobile
+#' Template for mobile:
+#' list(areas = [data.frame of strings],
+#'      rows_height = [vector of strings],
+#'      cols_width = [vector of strings])
+#'
+#' @return
+#' list(template = [string], area_names = [vector of strings])
+#'
+#' template - contains template that can be formatted with glue::glue() function
+#'
+#' area_names - contain all unique area names used in grid definition
+#'
+#' @details
+#' Example usage:
+#' \preformatted{
+#'   myGrid <- grid_template(default = list(
+#'     areas = rbind(
+#'       c("header", "header", "header"),
+#'       c("menu",   "main",   "right1"),
+#'       c("menu",   "main",   "right2")
+#'     ),
+#'     rows_height = c("50px", "auto", "100px"),
+#'     cols_width = c("100px", "2fr", "1fr")
+#'   ))
+#'
+#'   display_grid(myGrid)
+#' }
+#' Second example:
+#' \preformatted{
```

Comment:
For not exported function using `\preformatted` for code chunks is actually better but for the exported, we typically use `\examples`. Could you replace it?

Reply:
Done for `grid` and `grid_template` :heavy_check_mark:

Reply:
That's a bit weird, look how we do it elsewhere: https://github.com/Appsilon/shiny.semantic/blob/develop/R/rating.R#L15 and it works fine. Anyway, this is a minor thing. As long as is in correct section it's good enough.

Reply:
Ok, thanks for this link! I fixed it again, now it works with `@examples` only